File: source/ts/src/jyotish/lagna.ts
import { greenwichApparentSiderealDegrees } from '../astronomy/topocentric';
import { computeAyanamsa, dateToJulianDay } from '../astronomy/ayanamsa';
import { computeSunrise } from '../astronomy/sunrise';
import { getSiderealSunLongitude } from '../astronomy/sun';
import { meanObliquity } from './planets';
import { resolveNakshatraName, resolveMasaName } from '../i18n/resolver';
import { normalize360, degToRad } from '../utils/angle';
import { NAKSHATRA_SPAN, NAKSHATRA_PADA_SPAN, nakshatraOf } from '../utils/constants';
import { validateLocation, validateDate } from '../utils/validation';
import type { AyanamsaType, Language } from '../types/options';
import type { GeoLocation } from '../types/location';
import type { LagnaInfo, SripatiLagnaInfo } from '../types/jyotish';

/** Lagna (ascendant): the sidereal ecliptic longitude rising on the eastern horizon.
 *  Meeus, *Astronomical Algorithms* 2nd ed. eq. 13.6, less ayanamsa. */
export function computeLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType = 'lahiri',
  lang: Language = 'en',
): LagnaInfo {
  validateDate(birthDate);
  validateLocation(location);

  const lstDeg = normalize360(greenwichApparentSiderealDegrees(birthDate) + location.longitude);
  const θ = degToRad(lstDeg);

  const T = (dateToJulianDay(birthDate) - 2451545.0) / 36525.0;
  const ε = degToRad(meanObliquity(T));
  const φ = degToRad(location.latitude);

  const numerator = Math.cos(θ);
  const denominator = -Math.sin(ε) * Math.tan(φ) - Math.cos(ε) * Math.sin(θ);
  const tropicalLagna = normalize360((Math.atan2(numerator, denominator) * 180) / Math.PI);

  const ayanamsa = computeAyanamsa(birthDate, ayanamsaType);
  const siderealLongitude = normalize360(tropicalLagna - ayanamsa);

  const rashiIndex = Math.floor(siderealLongitude / 30);
  const degreeInRashi = siderealLongitude - rashiIndex * 30;
  const nakIdx = nakshatraOf(siderealLongitude);
  // Clamped at 0: a sub-ULP negative would floor pada to 0.
  const degreesInNakshatra = Math.max(siderealLongitude - nakIdx * NAKSHATRA_SPAN, 0);
  const pada = Math.min(4, Math.floor(degreesInNakshatra / NAKSHATRA_PADA_SPAN) + 1);

  return {
    siderealLongitude,
    rashi: { index: rashiIndex, name: resolveMasaName(rashiIndex, lang) },
    degreeInRashi,
    nakshatra: { index: nakIdx, name: resolveNakshatraName(nakIdx, lang) },
    pada,
  };
}

export function findSunriseBefore(date: Date, location: GeoLocation): Date {
  const back30h = new Date(date.getTime() - 30 * 3600_000);
  let candidate = computeSunrise(back30h, location);
  // 22 h hops never overshoot a solar day; the bound guards polar inputs.
  for (let i = 0; i < 4; i++) {
    const lookAhead = new Date(candidate.getTime() + 22 * 3600_000);
    const next = computeSunrise(lookAhead, location);
    if (next.getTime() > date.getTime()) return candidate;
    candidate = next;
  }
  return candidate;
}

/** @internal */
export const _findSunriseBeforeForTest = findSunriseBefore;

function buildLagnaInfo(siderealLongitude: number, lang: Language): LagnaInfo {
  const sid = normalize360(siderealLongitude);
  const rashiIndex = Math.floor(sid / 30);
  const degreeInRashi = sid - rashiIndex * 30;
  const nakIdx = nakshatraOf(sid);
  // Clamped at 0: a sub-ULP negative would floor pada to 0.
  const degreesInNakshatra = Math.max(sid - nakIdx * NAKSHATRA_SPAN, 0);
  const pada = Math.min(4, Math.floor(degreesInNakshatra / NAKSHATRA_PADA_SPAN) + 1);
  return {
    siderealLongitude: sid,
    rashi: { index: rashiIndex, name: resolveMasaName(rashiIndex, lang) },
    degreeInRashi,
    nakshatra: { index: nakIdx, name: resolveNakshatraName(nakIdx, lang) },
    pada,
  };
}

/** Hora Lagna: 30 deg/hour (one rashi per hour) from sunrise, BPHS Ch. 4. */
export function computeHoraLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType = 'lahiri',
  lang: Language = 'en',
): LagnaInfo {
  validateDate(birthDate);
  validateLocation(location);
  const sunrise = findSunriseBefore(birthDate, location);
  // BPHS Ch. 4: the base point is the Sun at sunrise, not the ascendant.
  const sunSidAtSunrise = getSiderealSunLongitude(sunrise, ayanamsaType);
  const hoursSinceSunrise = (birthDate.getTime() - sunrise.getTime()) / 3600_000;
  const horaLon = sunSidAtSunrise + hoursSinceSunrise * 30;
  return buildLagnaInfo(horaLon, lang);
}

/** Ghati Lagna: one rashi per ghatika (24 min) from sunrise, 75 deg/hour. */
export function computeGhatiLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType = 'lahiri',
  lang: Language = 'en',
): LagnaInfo {
  validateDate(birthDate);
  validateLocation(location);
  const sunrise = findSunriseBefore(birthDate, location);
  const sunSidAtSunrise = getSiderealSunLongitude(sunrise, ayanamsaType);
  const ghatikasSinceSunrise = (birthDate.getTime() - sunrise.getTime()) / (24 * 60 * 1000);
  const ghatiLon = sunSidAtSunrise + ghatikasSinceSunrise * 30;
  return buildLagnaInfo(ghatiLon, lang);
}

/** Bhava Lagna: one rashi per 5 ghatikas (2 h) from sunrise, 15 deg/hour. */
export function computeBhavaLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType = 'lahiri',
  lang: Language = 'en',
): LagnaInfo {
  validateDate(birthDate);
  validateLocation(location);
  const sunrise = findSunriseBefore(birthDate, location);
  const sunSidAtSunrise = getSiderealSunLongitude(sunrise, ayanamsaType);
  const hoursSinceSunrise = (birthDate.getTime() - sunrise.getTime()) / 3600_000;
  const bhavaLon = sunSidAtSunrise + hoursSinceSunrise * 15;
  return buildLagnaInfo(bhavaLon, lang);
}

/** Sripati Lagna: cusp 1 of the Sripati Paddhati, which trisects each of the four
 *  ASC/IC/DSC/MC quadrants into the 12 bhava madhyas (BPHS Ch. 5). */
export function computeSripatiLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType?: AyanamsaType,
  lang?: Language,
): LagnaInfo;
export function computeSripatiLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType | undefined,
  lang: Language | undefined,
  options: { includeCusps: true },
): SripatiLagnaInfo;
/** Explicit `{ includeCusps: false }`; without this overload such a call fails
 *  to compile. */
export function computeSripatiLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType | undefined,
  lang: Language | undefined,
  options?: { includeCusps?: false },
): LagnaInfo;
export function computeSripatiLagna(
  birthDate: Date,
  location: GeoLocation,
  ayanamsaType: AyanamsaType = 'lahiri',
  lang: Language = 'en',
  options?: { includeCusps?: boolean },
): LagnaInfo | SripatiLagnaInfo {
  const lagna = computeLagna(birthDate, location, ayanamsaType, lang);
  if (!options?.includeCusps) return lagna;

  // Sidereal MC, Meeus' tropical-MC formula; inlined rather than importing
  // computeBhava, which would create a bhava.ts -> lagna.ts cycle.
  const lstDeg = normalize360(greenwichApparentSiderealDegrees(birthDate) + location.longitude);
  const θ = degToRad(lstDeg);
  const T = (dateToJulianDay(birthDate) - 2451545.0) / 36525.0;
  const ε = degToRad(meanObliquity(T));
  const mcTropical = normalize360(
    (Math.atan2(Math.sin(θ), Math.cos(θ) * Math.cos(ε)) * 180) / Math.PI,
  );
  const ayanamsa = computeAyanamsa(birthDate, ayanamsaType);
  const mcSidereal = normalize360(mcTropical - ayanamsa);

  return { ...lagna, cusps: sripatiCusps(lagna.siderealLongitude, mcSidereal) };
}

function sripatiCusps(ascSidereal: number, mcSidereal: number): number[] {
  const ASC = normalize360(ascSidereal);
  const MC  = normalize360(mcSidereal);
  const IC  = normalize360(MC + 180);
  const DSC = normalize360(ASC + 180);

  const arc_q1 = normalize360(IC  - ASC);
  const arc_q2 = normalize360(DSC - IC);
  const arc_q3 = normalize360(MC  - DSC);
  const arc_q4 = normalize360(ASC + 360 - MC);

  return [
    ASC,
    normalize360(ASC + arc_q1 / 3),
    normalize360(ASC + 2 * arc_q1 / 3),
    IC,
    normalize360(IC  + arc_q2 / 3),
    normalize360(IC  + 2 * arc_q2 / 3),
    DSC,
    normalize360(DSC + arc_q3 / 3),
    normalize360(DSC + 2 * arc_q3 / 3),
    MC,
    normalize360(MC  + arc_q4 / 3),
    normalize360(MC  + 2 * arc_q4 / 3),
  ];
}

File: source/go/internal/jyotish/lagna.go
package jyotish

import (
	"math"

	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/i18n"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/jsnum"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang-ts/source/go/v5/internal/utils"
)

func resolveAyanamsa(a types.AyanamsaType) types.AyanamsaType {
	if a == "" {
		return types.Lahiri
	}
	return a
}

func resolveLang(l types.Language) types.Language {
	if l == "" {
		return types.LanguageEn
	}
	return l
}

func ComputeLagna(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.LagnaInfo, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return types.LagnaInfo{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return types.LagnaInfo{}, err
	}
	ayanamsaType = resolveAyanamsa(ayanamsaType)
	lang = resolveLang(lang)

	lstDeg := utils.Normalize360(astronomy.GreenwichApparentSiderealDegrees(ctx, birthMs) + location.Longitude)
	theta := utils.DegToRad(lstDeg)

	t := (astronomy.DateToJulianDay(birthMs) - 2451545.0) / 36525.0
	eps := utils.DegToRad(MeanObliquity(t))
	phi := utils.DegToRad(location.Latitude)

	numerator := math.Cos(theta)
	denominator := -float64(math.Sin(eps)*math.Tan(phi)) - float64(math.Cos(eps)*math.Sin(theta))
	tropicalLagna := utils.Normalize360((math.Atan2(numerator, denominator) * 180) / jsnum.PI)

	ayanamsa, err := astronomy.ComputeAyanamsa(birthMs, ayanamsaType)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	siderealLongitude := utils.Normalize360(tropicalLagna - ayanamsa)

	rashiIndex := int(math.Floor(siderealLongitude / 30))
	degreeInRashi := siderealLongitude - float64(float64(rashiIndex)*30)
	nakIdx := utils.NakshatraOf(siderealLongitude)
	degreesInNakshatra := math.Max(siderealLongitude-float64(float64(nakIdx)*utils.NakshatraSpan), 0)
	pada := int(math.Floor(degreesInNakshatra/utils.NakshatraPadaSpan)) + 1
	if pada > 4 {
		pada = 4
	}

	return types.LagnaInfo{
		SiderealLongitude: siderealLongitude,
		Rashi:             types.RashiInfo{Index: rashiIndex, Name: i18n.ResolveMasaName(rashiIndex, lang)},
		DegreeInRashi:     degreeInRashi,
		Nakshatra:         types.LagnaNakshatra{Index: nakIdx, Name: i18n.ResolveNakshatraName(nakIdx, lang)},
		Pada:              pada,
	}, nil
}

func findSunriseBefore(ctx *astronomy.EphemerisCtx, ms int64, location types.GeoLocation) (int64, error) {
	back30h := ms - 30*3600_000
	candidate, err := astronomy.ComputeSunrise(ctx, back30h, location, astronomy.DefaultRiseSetLimitDays)
	if err != nil {
		return 0, err
	}
	for i := 0; i < 4; i++ {
		lookAhead := candidate + 22*3600_000
		next, err := astronomy.ComputeSunrise(ctx, lookAhead, location, astronomy.DefaultRiseSetLimitDays)
		if err != nil {
			return 0, err
		}
		if next > ms {
			return candidate, nil
		}
		candidate = next
	}
	return candidate, nil
}

func buildLagnaInfo(siderealLongitude float64, lang types.Language) types.LagnaInfo {
	sid := utils.Normalize360(siderealLongitude)
	rashiIndex := int(math.Floor(sid / 30))
	degreeInRashi := sid - float64(float64(rashiIndex)*30)
	nakIdx := utils.NakshatraOf(sid)
	degreesInNakshatra := math.Max(sid-float64(float64(nakIdx)*utils.NakshatraSpan), 0)
	pada := int(math.Floor(degreesInNakshatra/utils.NakshatraPadaSpan)) + 1
	if pada > 4 {
		pada = 4
	}
	return types.LagnaInfo{
		SiderealLongitude: sid,
		Rashi:             types.RashiInfo{Index: rashiIndex, Name: i18n.ResolveMasaName(rashiIndex, lang)},
		DegreeInRashi:     degreeInRashi,
		Nakshatra:         types.LagnaNakshatra{Index: nakIdx, Name: i18n.ResolveNakshatraName(nakIdx, lang)},
		Pada:              pada,
	}
}

func ComputeHoraLagna(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.LagnaInfo, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return types.LagnaInfo{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return types.LagnaInfo{}, err
	}
	ayanamsaType = resolveAyanamsa(ayanamsaType)
	lang = resolveLang(lang)

	sunriseMs, err := findSunriseBefore(ctx, birthMs, location)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	sunSidAtSunrise, err := astronomy.GetSiderealSunLongitude(ctx, sunriseMs, ayanamsaType)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	hoursSinceSunrise := float64(birthMs-sunriseMs) / 3600_000 // int64 division would truncate
	horaLon := sunSidAtSunrise + float64(hoursSinceSunrise*30)
	return buildLagnaInfo(horaLon, lang), nil
}

func ComputeGhatiLagna(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.LagnaInfo, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return types.LagnaInfo{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return types.LagnaInfo{}, err
	}
	ayanamsaType = resolveAyanamsa(ayanamsaType)
	lang = resolveLang(lang)

	sunriseMs, err := findSunriseBefore(ctx, birthMs, location)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	sunSidAtSunrise, err := astronomy.GetSiderealSunLongitude(ctx, sunriseMs, ayanamsaType)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	ghatikasSinceSunrise := float64(birthMs-sunriseMs) / (24 * 60 * 1000)
	ghatiLon := sunSidAtSunrise + float64(ghatikasSinceSunrise*30)
	return buildLagnaInfo(ghatiLon, lang), nil
}

func ComputeBhavaLagna(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.LagnaInfo, error) {
	if err := utils.ValidateDate(birthMs); err != nil {
		return types.LagnaInfo{}, err
	}
	if err := utils.ValidateLocation(location); err != nil {
		return types.LagnaInfo{}, err
	}
	ayanamsaType = resolveAyanamsa(ayanamsaType)
	lang = resolveLang(lang)

	sunriseMs, err := findSunriseBefore(ctx, birthMs, location)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	sunSidAtSunrise, err := astronomy.GetSiderealSunLongitude(ctx, sunriseMs, ayanamsaType)
	if err != nil {
		return types.LagnaInfo{}, err
	}
	hoursSinceSunrise := float64(birthMs-sunriseMs) / 3600_000
	bhavaLon := sunSidAtSunrise + float64(hoursSinceSunrise*15)
	return buildLagnaInfo(bhavaLon, lang), nil
}

func ComputeSripatiLagna(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.LagnaInfo, error) {
	return ComputeLagna(ctx, birthMs, location, ayanamsaType, lang)
}

func ComputeSripatiLagnaWithCusps(
	ctx *astronomy.EphemerisCtx,
	birthMs int64,
	location types.GeoLocation,
	ayanamsaType types.AyanamsaType,
	lang types.Language,
) (types.SripatiLagnaInfo, error) {
	lagna, err := ComputeLagna(ctx, birthMs, location, ayanamsaType, lang)
	if err != nil {
		return types.SripatiLagnaInfo{}, err
	}
	ayanamsaType = resolveAyanamsa(ayanamsaType)

	lstDeg := utils.Normalize360(astronomy.GreenwichApparentSiderealDegrees(ctx, birthMs) + location.Longitude)
	theta := utils.DegToRad(lstDeg)
	t := (astronomy.DateToJulianDay(birthMs) - 2451545.0) / 36525.0
	eps := utils.DegToRad(MeanObliquity(t))
	mcTropical := utils.Normalize360(
		(math.Atan2(math.Sin(theta), math.Cos(theta)*math.Cos(eps)) * 180) / jsnum.PI,
	)
	ayanamsa, err := astronomy.ComputeAyanamsa(birthMs, ayanamsaType)
	if err != nil {
		return types.SripatiLagnaInfo{}, err
	}
	mcSidereal := utils.Normalize360(mcTropical - ayanamsa)

	return types.SripatiLagnaInfo{
		LagnaInfo: lagna,
		Cusps:     sripatiCusps(lagna.SiderealLongitude, mcSidereal),
	}, nil
}

func sripatiCusps(ascSidereal, mcSidereal float64) []float64 {
	asc := utils.Normalize360(ascSidereal)
	mc := utils.Normalize360(mcSidereal)
	ic := utils.Normalize360(mc + 180)
	dsc := utils.Normalize360(asc + 180)

	arcQ1 := utils.Normalize360(ic - asc)
	arcQ2 := utils.Normalize360(dsc - ic)
	arcQ3 := utils.Normalize360(mc - dsc)
	arcQ4 := utils.Normalize360(asc + 360 - mc)

	return []float64{
		asc,
		utils.Normalize360(asc + arcQ1/3),
		utils.Normalize360(asc + 2*arcQ1/3),
		ic,
		utils.Normalize360(ic + arcQ2/3),
		utils.Normalize360(ic + 2*arcQ2/3),
		dsc,
		utils.Normalize360(dsc + arcQ3/3),
		utils.Normalize360(dsc + 2*arcQ3/3),
		mc,
		utils.Normalize360(mc + arcQ4/3),
		utils.Normalize360(mc + 2*arcQ4/3),
	}
}
